heka/sandbox/filters/telemetry_latency.lua

```diff
@@ -0,0 +1,224 @@
+local PING_VERSION = "4"
+local NSPERHOUR = 60*60*1e9
+local NSPERDAY = 24*NSPERHOUR
+local SECPERHOUR = NSPERHOUR/1e9
+local MEDIAN = 1
+local PERCENTILE_75 = 2
+local PERCENTILE_99 = 3
+
+local seen_by_channel = {}
+local seen_history_by_channel = {}
+
+local creation_delay_by_channel = {}
+local creation_delay_history_by_channel = {}
+
+local creation_delay24h_by_channel = {}
+local creation_delay24h_history_by_channel = {}
+
+local creation_delay7d_by_channel = {}
+local creation_delay7d_history_by_channel = {}
+
+local rows = read_config("rows") or 1440
+local sec_per_row = read_config("sec_per_row") or 60
+
+local beginning_of_time = nil
+local last_subsessions_cleanup = nil
+
+local function log(message)
+    local dbg = {message}
+    inject_payload("txt", "debug", table.concat(dbg, "\n"))
+end
+
+local function get_channel_entry(t, channel, build)
+   local entry = t[channel]
+   if not entry then
+      entry = build()
+      t[channel] = entry
+   end
+   return entry
+end
+
+local function get_history(unit, metric_history_by_channel, channel)
+   local history = metric_history_by_channel[channel]
+   if not history then
+      history = circular_buffer.new(rows, 3, sec_per_row)
+      history:set_header(MEDIAN, "Median", unit, "none")
+      history:set_header(PERCENTILE_75, "75th percentile", unit, "none")
+      history:set_header(PERCENTILE_99, "99th percentile", unit, "none")
+      metric_history_by_channel[channel] = history
+   end
+   return history
+end
+
+local function process_client_metric(metric_by_channel, channel, client_id, value)
+   local metric = get_channel_entry(metric_by_channel, channel,
+                                    function() return {} end)
+   metric[client_id] = value
+end
+
+local function new_subsession_struct(rows)
+   return {list = {}, counter= circular_buffer.new(rows, 1, SECPERHOUR)}
+end
+
+local function process_subsession_metric(ts, metric_by_channel, hours, channel, value)
+   local metric = get_channel_entry(metric_by_channel, channel,
+                                    function() return new_subsession_struct(hours) end)
+   table.insert(metric.list, value)
+   metric.counter:add(ts, 1, 1)
+end
+
+function process_message ()
+   local sample_id = read_message("Fields[sampleId]")
+   local version = read_message("Fields[sourceVersion]")
+
+   if version == PING_VERSION and sample_id == 0 then
+      local ts = read_message("Timestamp")
+      local client_id = read_message("Fields[clientId]")
+      local creation_ts = read_message("Fields[creationTimestamp]") -- exists only in new "unified" pings
+      local channel = read_message("Fields[appUpdateChannel]") or "UNKNOWN"
+      local reason = read_message("Fields[reason]")
+      local ch_reason = channel .. ":" .. reason
+
+      process_client_metric(seen_by_channel, ch_reason, client_id, ts)
+      process_client_metric(creation_delay_by_channel, ch_reason, client_id, ts - creation_ts)
```

Comment:
The client metrics tables could be consolidated (it would save space and lookups).  The same for the subsession metrics.  If the size or speed is a problem.